### PR TITLE
ASR SSL Update

### PR DESCRIPTION
**Pull request by sam1373**
# What does this PR do ?

Update contrastive ssl for asr

**Collection**: ASR

# Changelog 
- add options to contrastive loss implementation
- add slightly different spec augmentation for ssl
- update corresponding configs for conformer, citrinet, contextnet ssl

**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

### CODE REVIEW

File: examples/asr/conf/ssl/citrinet/citrinet_ssl_1024.yaml [hunk 1]
```diff
@@ -402,22 +402,22 @@ model:
...
 
-
   decoder:
     _target_: nemo.collections.asr.modules.ConvASRDecoderReconstruction
     feat_in: ${model.model_defaults.enc_final}
     feat_hidden: 128
     feat_out: ${model.model_defaults.decoder_out_channels}
     stride_layers: 1
-
+    non_stride_layers: 0
```
Comment (titu1994) [inline examples/asr/conf/ssl/citrinet/citrinet_ssl_1024.yaml:411]: It would be nice to have comments across the values of this config to understand what are legal values and what they affect

File: nemo/collections/asr/losses/pt_losses/contrastive.py [hunk 2]
```diff
@@ -48,12 +48,13 @@ def __init__(
         proj_dim: int = 128,
         combine_time_steps: int = 1,
         num_negatives: int = 100,
-        quantized_targets: bool = True,
+        quantized_targets: bool = False,
```
Comment (titu1994) [inline nemo/collections/asr/losses/pt_losses/contrastive.py:51]: It is fine for now since we have not released pretrained checkpoints, however changing default like this will break older Nemo files. Consider that when designing / adding breaking changes

File: nemo/collections/asr/losses/pt_losses/contrastive.py [hunk 3]
```diff
@@ -116,13 +119,16 @@ def __init__(
             self.target_proj = nn.Linear(in_dim * combine_time_steps, proj_dim)
 
     def sample_negatives(self, y, num):
+        # y - T'x...
```
Comment (titu1994) [inline nemo/collections/asr/losses/pt_losses/contrastive.py:122]: ?

File: nemo/collections/asr/losses/pt_losses/contrastive.py [hunk 4]
```diff
@@ -116,13 +119,16 @@ def __init__(
...
 
         high = y.shape[0]
-        with torch.no_grad():
-            neg_idxs = torch.randint(low=0, high=high - 1, size=(self.num_negatives * num,))
+        neg_idxs = torch.multinomial(torch.ones(num, high), self.num_negatives).to(device=y.device)
```
Comment (titu1994) [inline nemo/collections/asr/losses/pt_losses/contrastive.py:125]: For efficiency, use device directly in the ones() call

**Review by titu1994 — APPROVED** [review]
Looks good, minor suggestions

**Review by titu1994 — APPROVED** [review]
Looks great !
